### PR TITLE
feat(consent)!: migrate to UMP SDK

### LINKED ISSUE #38: Personalized Ad Consent SDK deprecated in favor of the User Messaging Platform SDK
The Personalized Ad Consent SDK is no longer being maintained. It has been deprecated in favor of the User Messaging Platform SDK , which is IAB compliant.

See: https://developers.google.com/admob/ump/ios/quick-start

### LINKED ISSUE #43: adconsent setDebugGeography dosen't show effect using emulator
am using emulator and setDebugGeography isn't working

### PR DESCRIPTION
**Pull request by birgernass**
### Description

This PR replaces the legacy consent library with the new UMP SDK v2.
It is no longer possible to request the ad providers or to show a custom form.

There are also some platform inconsistencies I tried to resolve.
Please let me know if you see better ways to compromise:
* The consent status integer values are different on iOS and Android. I decided to return the respective string values instead.
* There is no equivalent for `UMPConsentInformation.sharedInstance.formStatus` on Android. I decided to return `isConsentFormAvailable` on both platforms, as the formStatus doesn't add much value as far as I'm concerned.

### Related issues

Fixes #38
Closes #43
Related to #64

### Release Summary

* changed `requestInfoUpdate` (new parameter & return type)
* changed `AdsConsentStatus` (string instead of integer)
* removed `showForm` (new parameter & return type)
* removed `addTestDevices` (see requestInfoUpdate)
* removed `setDebugGeography` (see requestInfoUpdate)
* removed `setTagForUnderAgeOfConsent` (see requestInfoUpdate)
* removed `getAdProviders`
* removed `getStatus`
* removed `setStatus`
* added `reset` method to reset state during debugging
* added optional `user_tracking_usage_description` key to `app.json` for ATT support

A library like [@iabtcf/core](https://github.com/InteractiveAdvertisingBureau/iabtcf-es/tree/master/modules/core#iabtcfcore) can be used to parse the TCF string and evaluate the users consent choices.

### Checklist

- I read the [Contributor Guide](https://github.com/invertase/react-native-google-mobile-ads/blob/main/CONTRIBUTING.md)
  and followed the process outlined there for submitting PRs.
  - [X] Yes 🔥 
- My change supports the following platforms;
  - [X] `Android`
  - [X] `iOS`
- My change includes tests;
  - [X] `e2e` tests added or updated in `__tests__e2e__`
  - [X] `jest` tests added or updated in `__tests__`
- [X] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [X] Yes
  - [ ] No

### Test Plan

<p align="center">
Screenshot 2022-02-25 at 22 16 07<br>
(The io.invertase.googleads package has since been updated.)
</p>
---

Think `react-native-google-mobile-ads` is great? Please consider supporting the project with any of the below:

- 👉 Star this repo on GitHub ⭐️
- 👉 Follow [`Invertase`](https://twitter.com/invertaseio) on Twitter

### CODE REVIEW

File: ios/RNGoogleMobileAds/RNGoogleMobileAdsConsentModule.m [hunk 1]
```diff
@@ -18,8 +18,17 @@
...
+#import "RCTBridgeModule.h"
 #import "RNGoogleMobileAdsConsentModule.h"
 #import "common/RNSharedUtils.h"
+#if __has_include(<React/RCTConvert.h>)
+#import <React/RCTConvert.h>
+#elif __has_include("RCTConvert.h")
+#import "RCTConvert.h"
+#else
+#import "React/RCTConvert.h"
+#endif
```
Comment (mikehardy) [inline ios/RNGoogleMobileAds/RNGoogleMobileAdsConsentModule.m:30]: I don't believe this has been necessary for a long time - I think they can just all be one-liner `#import <React/RCTConvert.h>` style headers now?
I seem to recall a PR taking this conditional import stuff *out* of some other module recently even, where it was commented that it was an artifact of some long long gone react-native version

File: ios_config.sh [hunk 2]
```diff
@@ -102,6 +102,14 @@ if [[ ${_SEARCH_RESULT} ]]; then
...
     _PLIST_ENTRY_VALUES+=("$_IOS_APP_ID")
   fi
+
+    # config.user_tracking_usage_description
+  _USER_TRACKING_USAGE_DESCRIPTION=$(getJsonKeyValue "$_JSON_OUTPUT_RAW" "user_tracking_usage_description")
+  if [[ $_USER_TRACKING_USAGE_DESCRIPTION ]]; then
+    _PLIST_ENTRY_KEYS+=("NSUserTrackingUsageDescription")
+    _PLIST_ENTRY_TYPES+=("string")
+    _PLIST_ENTRY_VALUES+=("$_USER_TRACKING_USAGE_DESCRIPTION")
+  fi
```
Comment (mikehardy) [inline ios_config.sh:112]: @Salakar did you have any luck with the cocoapods add-plist functionality? I wonder if the reason we weren't using it before was the conditional JSON stuff, but ruby could probably handle it if you had a good result and maybe we could remove all this ios_config.sh machinery

File: android/src/main/java/io/invertase/googlemobileads/ReactNativeGoogleMobileAdsConsentModule.java [hunk 3]
```diff
@@ -21,206 +21,134 @@
...
-        new ConsentInfoUpdateListener() {
-          @Override
-          public void onConsentInfoUpdated(ConsentStatus consentStatus) {
+  public void requestInfoUpdate(final ReadableMap options, final Promise promise) {
+    try {
+      ConsentRequestParameters.Builder paramsBuilder = new ConsentRequestParameters.Builder();
+      ConsentDebugSettings.Builder debugSettingsBuilder =
+          new ConsentDebugSettings.Builder(getApplicationContext());
+
+      if (options.hasKey("testDeviceIdentifiers")) {
```
Comment (mikehardy) [inline android/src/main/java/io/invertase/googlemobileads/ReactNativeGoogleMobileAdsConsentModule.java:64]: This is a nit-pick, but we're dereferencing options here without noting it is `@Nonnull` as a parameter annotation or testing it for null. Further on I can see at the typescript level you're verifying it is an object so this is probably safe and just a code style thing. I'd go with a `@Nonnull` annotation from androidx.annotations normally then, or leave it if that doesn't work

File: docs/european-user-consent.mdx [hunk 4]
```diff
@@ -54,156 +45,76 @@ npx react-native run-ios
...
+Once set, rebuild your application.
 
-Your own publisher ID can be found on the [Google AdMob dashboard](https://apps.admob.com/v2/settings) under "Settings > Account Information":
+### Requesting consent information
 
-![Publisher ID](https://prismic-io.s3.amazonaws.com/invertase%2Fbcb7e7d6-2bb3-4e8c-8e6e-be99d6bc4f56_new+project+%2816%29.jpg)
+It is recommended you request consent information each time your application starts to determine if the consent modal should be shown, e.g. due to provider changes.
 
-To request consent, call the method as early as possible within your app before presenting any ads:
+The `AdsConsent` helper provides a promised based method to return the users consent status called `requestInfoUpdate`.
```
Comment (mikehardy) [inline docs/european-user-consent.mdx:69]: tiny grammar nitpick

```suggestion
The `AdsConsent` helper provides a promise based method to return the users consent status called `requestInfoUpdate`.
```

File: docs/european-user-consent.mdx [hunk 5]
```diff
@@ -54,156 +45,76 @@ npx react-native run-ios
...
-The result of the method returns an `AdsConsentInfo` interface, which provides information about the users status and location:
-
-- **status**: The status can be one of 3 outcomes:
-  - `UNKNOWN`: The user has not yet given consent, or has been invalidated since the last time they gave consent.
-  - `NON_PERSONALIZED`: The user gave consent to see non-personalized ads only.
-  - `PERSONALIZED`: The user gave consent to see personalized ads.
-- **isRequestLocationInEeaOrUnknown**: A boolean value. If `true` the users location is within the EEA or is unknown.
-
-The combination of status and location allows you to handle the next steps for requesting consent, if required:
+The result of the method returns an `AdsConsentInfo` interface, which provides information about consent form availability and the users consent status:
```
Comment (mikehardy) [inline docs/european-user-consent.mdx:77]:
```suggestion
The method returns an `AdsConsentInfo` interface, which provides information about consent form availability and the users consent status:
```

File: docs/european-user-consent.mdx [hunk 6]
```diff
@@ -54,156 +45,76 @@ npx react-native run-ios
...
 
 When developing the consent flow, the behavior of the `AdsConsent` responses may not be reliable due to the environment
 (e.g. using an emulator vs real device). It is possible to set a debug location to test the various responses from the
-Consent SDK.
+UMP SDK.
 
-If using a real device, ensure you whitelist it using the device ID, which can be obtained from native logs or using a library
-such as [react-native-device-info](https://github.com/react-native-community/react-native-device-info). Once found,
-call the `addTestDevice(deviceId)` method.
+If using a real device, ensure you whitelist it by passing the device ID, which can be obtained from native logs or using a library
```
Comment (mikehardy) [inline docs/european-user-consent.mdx:112]:
```suggestion
If using a real device, ensure you add it to the list of allowed devices by passing the device ID, which can be obtained from native logs or using a library
```

File: docs/european-user-consent.mdx [hunk 7]
```diff
@@ -214,45 +125,16 @@ For example:
...
-
-### Forwarding the consent status to ads
-
-Assuming the user is within the EEA and has provided consent, their status needs to be forwarded to every ad request we
-make in our application.
-
-> If the user is within the EEA and has not given consent, do not display AdMob ads (even non-personalized).
-
-Taking a Rewarded Video as an example, we can apply the users consent when our ad is loaded via the `RequestOptions`. For example:
+It is also possible to reset the UMP state for test devices. This does not reset the ATT state though.
```
Comment (mikehardy) [inline docs/european-user-consent.mdx:134]: > This does not reset the ATT state though

Hmm - any pointers for inexperienced devs here? How does one reset the ATT state? Maybe a link to an apple doc or similar...

File: src/types/AdsConsent.interface.ts [hunk 8]
```diff
@@ -10,261 +10,87 @@ import { AdsConsentStatus } from '../AdsConsentStatus';
...
 
   /**
    * Set to `true` to provide the option for the user to accept being shown personalized ads, defaults to `false`.
    */
-  withPersonalizedAds?: boolean;
+  tagForUnderAgeOfConsent?: boolean;
 
   /**
-   * Set to `true` to provide the option for the user to accept being shown non-personalized ads, defaults to `false`.
+   * An array of test device IDs to whitelist.
```
Comment (mikehardy) [inline src/types/AdsConsent.interface.ts:91]:
```suggestion
   * An array of test device IDs to allow.
```

**Review by mikehardy — CHANGES_REQUESTED** [review]
I added a bunch of nit-picky comments but in general this is an amazing advance for the module and none of my comments reflect anything serious.
I'll defer to @dylancom on whatever he has to say but I'm pretty much +1 on all of it.
Only thought I have is it might pay off (in terms of lower support via issues logged) to do a quick migration document indicating the path someone using the old style might take to adopt this one. Does that make sense, and if so is it possible - even if pretty rough / terse?

**Comment by birgernass** [comment]
> I added a bunch of nit-picky comments but in general this is an amazing advance for the module and none of my comments reflect anything serious. I'll defer to @dylancom on whatever he has to say but I'm pretty much +1 on all of it. Only thought I have is it might pay off (in terms of lower support via issues logged) to do a quick migration document indicating the path someone using the old style might take to adopt this one. Does that make sense, and if so is it possible - even if pretty rough / terse?

Happy to hear! And thank you for the suggestions. Also makes a lot of sense to add a bit more context for the migration. I'll write something up. Should it be part of the web docs or just added to the change log or what did you have in mind?

**Comment by mikehardy** [comment]
> > I added a bunch of nit-picky comments but in general this is an amazing advance for the module and none of my comments reflect anything serious. I'll defer to @dylancom on whatever he has to say but I'm pretty much +1 on all of it. Only thought I have is it might pay off (in terms of lower support via issues logged) to do a quick migration document indicating the path someone using the old style might take to adopt this one. Does that make sense, and if so is it possible - even if pretty rough / terse?
> 
> Happy to hear! And thank you for the suggestions. Also makes a lot of sense to add a bit more context for the migration. I'll write something up. Should it be part of the web docs or just added to the change log or what did you have in mind?

I think it makes sense as a file in the docs area that goes to the web docs. That way it can also be easily changed/improved by anyone that tries it later, changelogs end up being kind of immutable / irritating to change

**Review by mikehardy — APPROVED** [review]
Thanks for the quick turnaround!
I'm curious if Mike Diarmid (Salakar) has input on the ios_config thing but that is legitimately a different PR really if he does, so no need to block on it.
I'm +1 on this then but will defer to Dylan in case he sees something I missed (I always miss something)

**Comment by dylancom** [comment]
Awesome work! Will have time tomorrow to check this out.

**Comment by mikehardy** [comment]
I read the migration doc - I'm still +1 or now I guess I'm +2 because I think that will help a lot, and now if/when there is confusion people can post PRs to that doc as they work through it in case anything isn't clear. Gives us some place to point though as the inevitable issues will come in with "help what happened to AdConsent !?!"  ;-)

**Review by dylancom — APPROVED** [review]

**Comment by dylancom** [comment]
Looking good to me!

**Comment by mikehardy** [comment]
I had contemplated batching a Swift/Kotlin migration at the same time as this as one major version but I prefer to keep changes more incremental with regard to version numbers if possible. I really don't mind bumping major version but I like the versions themselves to be focused and hopefully stable. Mixing in a code migration at the same time as this release would go against that I think, so...thinking of just releasing this. Any objections? 

Maybe we can do one PR first to add this functionality to the example, and/or someone could test integrate it in their app to give it some semi-production use prior to full release? yarn lets you rely on the commit hash fa88240d937499b4a570c40914f06054dfa31b89 and if you add a prepare phase where you cd into the directory and yarn / yarn build it should be possible to depend on this from source at the commit point (that's a supported feature - building from source via commit ref - please holler if it does not work!)

**Comment by birgernass** [comment]
@mikehardy I prepared a [test case](https://github.com/birgernass/react-native-google-mobile-ads/commit/a9c03aeaebad233d96b77585c7b3df9abbc16149) that works with a real app ID, but doesn't work with the test app ID as the consent form needs to be configured in the AdMob backend.

I also can't run a production test at this point as we depend on a custom Ad Manager wrapper and I don't want to ship another library just to test it.

**Pull request merged**